syntax = "proto3";
import "type.proto";

service RendezvousNode {
    rpc get_request (NodeGetRequest) returns (NodeGetReply) {};  
    rpc get_objects (NodeEmpty) returns (stream NodeGetObjectsReply) {};   
    rpc get_replicas (NodeEmpty) returns (stream NodeGetReplicasReply) {};  
    rpc hash_value_for_key (NodeHashValueForRequest) returns (NodeHashValueForReply) {};
    rpc send_item_to_new_node (NodeSendItemToNewNodeRequest) returns (NodeEmpty) {};
    rpc remove_all (NodeEmpty) returns (NodeEmpty);
}

message NodeGetRequest {
    GetRequestType type = 1;
    string key = 2;
    repeated string values = 3;
    int32 replica_number = 4;
}

message NodeGetReply {
    repeated string values = 1;
}

message NodeGetObjectsReply {
    optional string key = 1;
    repeated string values = 2;
}

message NodeGetReplicasReply {
    optional string key = 1;
    repeated string values = 2;
}

message NodeHashValueForRequest {
    string key = 1;

}
message NodeHashValueForReply {
    float hashValue = 1;
}

message NodeSendItemToNewNodeRequest {
    string ip_address = 1;
}

message NodeEmpty {
}